os/binfmt/libelf/libelf_cache.c

```diff
@@ -0,0 +1,494 @@
+			if (ptr->no_requests_for_block >= max_accessed_count) {
+				max_accessed_count = ptr->no_requests_for_block;
+				most_accessed = ptr;
+			}
+
+			/* Always attach ptr at tail */
+			temp = tail;
+			temp->next = ptr;
+			ptr->prev = temp;
+			ptr->next = NULL;
+			tail = ptr;
+		}
+	}
+
+	return blockcache_index;
+}
+
+/****************************************************************************
+ * Name: libelf_cache_read
+ *
+ * Description:
+ *   Read bytes from the elf file using 'offset' and 'readsize' info
+ *   provided for cached buffer.  The data is read into 'buffer'. Offset
+ *   value here is offset from start of elf binary (excluding binary
+ *   header).
+ *
+ * Returned Value:
+ *   Number of bytes read into buffer on Success
+ *   Negative value on failure
+ ****************************************************************************/
+int libelf_cache_read(int filfd, uint16_t binary_header_size, FAR uint8_t *buffer, size_t readsize, off_t offset)
+{
+	int first_block;
+	int last_block;
+	int no_blocks;
+	int block_number;		/* Block number in an ELF file */
+	int actual_offset;		/* Offset from start of ELF file */
+	int block_size_to_write;	/* Size to write into buffer from cached block */
+	int buffer_pos;			/* Position in buffer to start writing from */
+	int blocksize;			/* Blocksize used for compressing the binary */
+	unsigned int blockcache_index;	/* Which blockcache element has needed ELF data for read */
+
+	binfo("filfd: %d readsize: %d offset: %d\n", filfd, readsize, offset);
+
+	/* Setting first block, end block and number of blocks to read */
+	blocksize = cache_blocks_size;
+	libelf_cache_blocks_to_read(&first_block, &last_block, &no_blocks, offset, readsize);
+	if (first_block < 0 || no_blocks < 0) {
+		berr("Incorrect first_block, no_blocks info\n");
+		buffer_pos = ERROR;
+		goto error_compress_read;
+	}
+
+	block_number = first_block;
+	buffer_pos = 0;
+	/* Actual Offset in ELF file is same as Offset passed to this function */
+	actual_offset = offset;
+
+	/* Reading and decompressing blocks from first_block to last_block. Then writing to buffer. */
+	for (; block_number < first_block + no_blocks; block_number++) {
+
+		/* Update blockcache list and get data into one of the blockcache elements */
+		blockcache_index = libelf_cache_update_blockcache_list(block_number, filfd, binary_header_size);
+		if (blockcache_index >= number_blocks_caching) {
+			buffer_pos = ERROR;
+			goto error_compress_read;
+		}
+
+		if (block_number == first_block) {
+			/*
+			 * Read appropriate number of bytes to buffer from this block using readsize and offset info.
+			 * If start_offset + readsize < end_offset, write from start_offset to start_offset + readsize.
+			 * Otherwise, write from start_offset to end_offset into buffer.
+			 */
+			block_size_to_write = ((block_number + 1) * blocksize - 1 > actual_offset + readsize - 1 ? readsize : (block_number + 1) * blocksize - actual_offset);
```

Comment:
Please provide more brackets for better understanding of calculation & code readability.